docs/profiling.md

```diff
@@ -1,12 +1,133 @@
+    shell, and run the short program with `%run` after starting the capture in
+    the next step. Another option is to start the profiler server at the
+    beginning of the program and use `time.sleep()` to give you enough time to
+    start the capture.<br /><br />
+
+1. Open <http://localhost:6006/#profile>, and click the "CAPTURE PROFILE" button
+   in the upper left. Enter "localhost:9999" as the profile service URL (this is
+   the address of the profiler server you started in the previous step). Enter
+   the number of milliseconds you'd like to profile for, and click "CAPTURE".<br
+   /><br />
+
+1. If the code you'd like to profile isn't already running (e.g. if you started
+   the profiler server in a Python shell), run it while the capture is
+   running.<br /><br />
+
+1. After the capture finishes, TensorBoard should automatically refresh. (Not
+   all of the TensorBoard profiling features are hooked up with JAX, so it may
+   initially look like nothing was captured.) On the left under "Tools", select
+   "trace_viewer".
+
+   You should now see a timeline of the execution. You can use the WASD keys to
+   navigate the trace, and click or drag to select events to see more details at
+   the bottom. See [these TensorFlow
+   docs](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras#use_the_tensorflow_profiler_to_profile_model_training_performance)
+   for more details on using the trace viewer.<br /><br />
+
+1. By default, the events in the trace viewer are mostly low-level internal JAX
+   functions. You can add your own events and functions by using
+   {func}`jax.profiler.TraceContext` and {func}`jax.profiler.trace_function` in
+   your code and capturing a new profile.
+
+### Troubleshooting
+
+#### GPU profiling
+
+Programs running on GPU should produce traces for the GPU streams near the top
+of the trace viewer. If you're only seeing the host traces, check your program
+logs and/or output for the following error messages.
+
+**If you get an error like: `Could not load dynamic library 'libcupti.so.10.1'`**<br />
+Full error:
+```
+W external/org_tensorflow/tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'libcupti.so.10.1'; dlerror: libcupti.so.10.1: cannot open shared object file: No such file or directory
+2020-06-12 13:19:59.822799: E external/org_tensorflow/tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1422] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
+```
+
+Add the path to `libcupti.so` to the environment variable `LD_LIBRARY_PATH`.
+(Try `locate libcupti.so` to find the path.) For example:
+```shell
+export LD_LIBRARY_PATH=/usr/local/cuda-10.1/extras/CUPTI/lib64/:$LD_LIBRARY_PATH
+```
+
+**If you get an error like: `failed with error CUPTI_ERROR_INSUFFICIENT_PRIVILEGES`**<br />
```

Comment:
Perhaps link to NVidia's documentation?

https://developer.nvidia.com/nvidia-development-tools-solutions-err-nvgpuctrperm-cupti

Reply:
Oh nice, I didn't see this actually.